File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/create/RoomCreateViewModel.java
package io.agora.scene.ktv.create;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import java.util.List;

import io.agora.scene.base.utils.ToastUtils;
import io.agora.scene.ktv.service.KTVCreateRoomInputModel;
import io.agora.scene.ktv.service.KTVCreateRoomOutputModel;
import io.agora.scene.ktv.service.KTVJoinRoomInputModel;
import io.agora.scene.ktv.service.KTVJoinRoomOutputModel;
import io.agora.scene.ktv.service.KTVServiceProtocol;
import io.agora.scene.ktv.service.VLRoomListModel;

public class RoomCreateViewModel extends AndroidViewModel {
    private final KTVServiceProtocol ktvServiceProtocol = KTVServiceProtocol.Companion.getImplInstance();

    public final MutableLiveData<List<VLRoomListModel>> roomModelList = new MutableLiveData<>();
    public final MutableLiveData<KTVJoinRoomOutputModel> joinRoomResult = new MutableLiveData<>();
    public final MutableLiveData<KTVCreateRoomOutputModel> createRoomResult = new MutableLiveData<>();

    public RoomCreateViewModel(@NonNull Application application) {
        super(application);
    }

    /**
     * 加载房间列表
     */
    public void loadRooms() {
        ktvServiceProtocol.getRoomList((e, vlRoomListModels) -> {
            if (e != null) {
                ToastUtils.showToast(e.getMessage());
                roomModelList.postValue(null);
                return null;
            }
            roomModelList.postValue(vlRoomListModels);
            return null;
        });
    }

    public void createRoom(int isPrivate,
                           String name, String password,
                           String userNo, String icon) {
        ktvServiceProtocol.createRoomWithInput(new KTVCreateRoomInputModel(
                icon, isPrivate, name, password, userNo, "", ""
        ), (e, ktvCreateRoomOutputModel) -> {
            if (e == null && ktvCreateRoomOutputModel != null) {
                // success
                createRoomResult.postValue(ktvCreateRoomOutputModel);
            } else {
                // failed
                if (e != null) {
                    ToastUtils.showToast(e.getMessage());
                }
                createRoomResult.postValue(null);
            }
            return null;
        });
    }

    public void joinRoom(String roomNo, String password) {
        ktvServiceProtocol.joinRoomWithInput(new KTVJoinRoomInputModel(roomNo, password),
                (e, ktvJoinRoomOutputModel) -> {
                    if (e == null && ktvJoinRoomOutputModel != null) {
                        // success
                        joinRoomResult.postValue(ktvJoinRoomOutputModel);
                    } else {
                        // failed
                        if (e != null) {
                            ToastUtils.showToast(e.getMessage());
                        }
                        joinRoomResult.postValue(null);
                    }
                    return null;
                });
    }

}

File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/create/RoomCreateViewModel.kt
package io.agora.scene.ktv.create

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import io.agora.rtmsyncmanager.model.AUIRoomInfo
import io.agora.scene.ktv.R
import io.agora.scene.ktv.service.KTVServiceProtocol.Companion.getImplInstance
import io.agora.scene.widget.toast.CustomToast

/**
 * The type Room create view model.
 */
class RoomCreateViewModel
/**
 * Instantiates a new Room create view model.
 *
 * @param application the application
 */
    (application: Application) : AndroidViewModel(application) {
    private val ktvServiceProtocol = getImplInstance()

    /**
     * The Room model list.
     */
    val roomModelList = MutableLiveData<List<AUIRoomInfo>?>()

    /**
     * The Join room result.
     */
    val roomInfoLiveData = MutableLiveData<AUIRoomInfo?>()

    /**
     * 加载房间列表
     */
    fun loadRooms() {
        ktvServiceProtocol.getRoomList { error, vlRoomListModels ->
            roomModelList.postValue(vlRoomListModels)
            error?.message?.let {
                CustomToast.show(it)
            }
        }
    }

    /**
     * Create room.
     *
     * @param name      the name
     * @param password  the password
     */
    fun createRoom(name: String, password: String) {
        ktvServiceProtocol.createRoom(name, password) { err, roomInfo ->
            if (err == null && roomInfo != null) {
                roomInfoLiveData.postValue(roomInfo)
            } else {
                roomInfoLiveData.postValue(null)
                CustomToast.show(R.string.ktv_create_room_failed, err?.message ?: "")
            }
        }
    }

    /**
     * Join room.
     *
     * @param roomNo   the room no
     * @param password the password
     */
    fun joinRoom(roomInfo: AUIRoomInfo, password: String?) {
        ktvServiceProtocol.joinRoom(roomInfo.roomId, password) { err ->
            if (err == null) { // success
                roomInfoLiveData.postValue(roomInfo)
            } else {
                roomInfoLiveData.postValue(null)
                CustomToast.show(R.string.ktv_join_room_failed, err.message ?: "")
            }
        }
    }
}
